File: 문제풀이/chap4-5/prob4-5-8.py
import heapq

# 입력
K = int(input())

# 인접리스트 작성 → 그래프 엣지 추가
G = [ list() for i in range(K) ]
for i in range(K):
	for j in range(10):
		if i == 0 and j == 0:
			continue
		G[i].append(((i * 10 + j) % K, j))

# 다익스트라 알고리즘: 배열 초기화 등
dist = [ 10 ** 10 ] * K
used = [ False ] * K
Q = list()
heapq.heappush(Q, (0, 0))  # 여기에서 dist[0] = 0로는 되지 않으므로 주의

# 다익스트라 알고리즘: 우선순위 큐 활용
while len(Q) >= 1:
	pos = heapq.heappop(Q)[1]
	if used[pos] == True:
		continue
	used[pos] = True
	for i in G[pos]:
		to = i[0]
		cost = dist[pos] + i[1]
		if pos == 0:
			cost = i[1]  # 정점 0의 경우는 제외
		if dist[to] > cost:
			dist[to] = cost
			heapq.heappush(Q, (dist[to], to))

# 답 출력
print(dist[0])

File: 문제풀이/chap4-5/prob4-5-8.cpp
#include <bits/stdc++.h>
using namespace std;
 
int K, dist[1 << 18];
bool used[1 << 18];
vector<pair<int, int>> G[1 << 18];
priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> Q;
 
// 다익스트라 알고리즘
void dijkstra() {
	// 배열 초기화 등
	for (int i = 0; i < K; i++) dist[i] = (1 << 30);
	for (int i = 0; i < K; i++) used[i] = false;
	Q.push(make_pair(0, 0)); // 여기에서 dist[0] = 0로는 되지 않으므로 주의

	// 큐 변경
	while (!Q.empty()) {
		int pos = Q.top().second; Q.pop();
		if (used[pos] == true) continue;
		used[pos] = true;
		for (pair<int, int> i : G[pos]) {
			int to = i.first, cost = dist[pos] + i.second;
			if (pos == 0) cost = i.second; // 정점 0의 경우는 제외
			if (dist[to] > cost) {
				dist[to] = cost;
				Q.push(make_pair(dist[to], to));
			}
		}
	}
}
 
int main() {
	// 입력
	cin >> K;
	
	// 그래프 엣지 추가
	for (int i = 0; i < K; i++) {
		for (int j = 0; j < 10; j++) {
			if (i == 0 && j == 0) continue;
			G[i].push_back(make_pair((i * 10 + j) % K, j));
		}
	}
	
	// 다익스트라 알고리즘/출력
	dijkstra();
	cout << dist[0] << endl;
	return 0;
}